File: 0414/큰 수 만들기.java
class Solution {
    
    static char[] answer;
    
    public String solution(String number, int k) {
        char[] chars = number.toCharArray();
        
        int cnt = number.length() - k;
        answer = new char[cnt];
        
        int start = 0; // 선택 시작 index
        for (int i=1; i<=cnt; i++) { // i번째 숫자 선택 과정
            char max = '0'; // 선택 범위 중 가장 큰 수
            for (int j=start; j<number.length() - (cnt - i); j++) {
                if (chars[j] > max) {
                    max = chars[j];
                    start = j + 1;
                }
            }
            answer[i - 1] = max;
        }
        
        return new String(answer);
    }
}

File: 0414/큰 수 만들기.py
from collections import deque

def solution(number, k):
    nums = deque([int(e) for e in number])
    stack = [nums.popleft()]

    cnt = 0
    
    while nums:
        x = stack[-1]
        y = nums.popleft()

        while x < y and stack and cnt != k:
            stack.pop()
            cnt += 1
            if stack:
                x = stack[-1]
                
        stack.append(y)

    if cnt == k:
        answer = ''.join(list(map(str, stack)))
    else:
        answer = ''.join(list(map(str, stack[:-(k - cnt)])))
        
    return answer
